packages/@aws-cdk/aws-lambda/lib/function.ts

```diff
@@ -705,6 +705,23 @@ export class Function extends FunctionBase {
       if (config.dependency) {
         this.node.addDependency(...config.dependency);
       }
+      // There could be a race if the Lambda is used in a CustomResource. It is possible for the Lambda to
+      // fail to attach to a given FileSystem if we do not have a dependency on the SecurityGroup ingress/egress
+      // rules that were created between this Lambda's SG & the Filesystem SG.
+      this.connections.securityGroups.forEach(sg => {
+        sg.node.findAll().forEach(child => {
+          if (child instanceof CfnResource && child.cfnResourceType === 'AWS::EC2::SecurityGroupEgress') {
+            resource.node.addDependency(child);
+          }
+        });
+      });
+      config.connections?.securityGroups.forEach(sg => {
+        sg.node.findAll().forEach(child => {
+          if (child instanceof CfnResource && child.cfnResourceType === 'AWS::EC2::SecurityGroupIngress') {
+            resource.node.addDependency(child);
+          }
+        });
+      });
```

Comment:
Why not just add a dependency to the `filesystem` construct itself? 

```ts
resource.node.addDependency(props.filesystem);
```

This should build a dependency transitively on all its security groups.

Further, that would also ensure that the EFS Filesystem is created before the Lambda Function as well. 
I believe if these two got created out of order, you will hit the same issue you're solving but for the Filesystem. Am I correct?

Reply:
I have gotten into trouble with creating a dependency on an entire L2 construct tree in the past. Specifically with security groups, and particularly with cross-stack definitions. It's very easy to get into circular references when shotgunning the dependencies, so I find it better to be surgical about it.